### PR TITLE
#164069234 Search articles by author, tags, favourited and keywords

### PR DESCRIPTION
**Pull request by BagzieGracious**
#### What does this PR do?
- Add search functionality  and custom filtering based on parameters
#### Description of Task to be completed?
- Ensure user can filter articles by author name
- Ensure user can filter articles by tag name
- Ensure user can filter articles by who favourited the articles
#### How should this be manually tested?
- Clone the [repo](https://github.com/andela/Ah-backend-aquaman), cd into `Ah-backend-aquaman` and checkout the branch `ft-search-article-164069234`
    -  pip install -r requirements.txt
    -  python manage.py migrate
    -  python manage.py runserver
    -  create an account, verify your email, login and create articles
    -  use postman to test the search endpoint below:-
        - `http://127.0.0.1:8000/api/articles/?author=jake`
        - `http://127.0.0.1:8000/api/articles/?keyword=soccer`
        - `http://127.0.0.1:8000/api/articles/?tag=AngularJS`
        - `http://127.0.0.1:8000/api/articles/?favorited=joel`
#### Any background context you want to provide?
- N/A
#### What are the relevant pivotal tracker stories?
- [#164069234](https://www.pivotaltracker.com/story/show/164069234)

#### Screenshots (if appropriate)
* search by keyword
Screenshot 2019-03-18 at 16 10 09

* search by tag name
Screenshot 2019-03-18 at 16 10 44

### CODE REVIEW

File: authors/apps/articles/tests/test_search.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+class TestSearchArticle(ArticlesBaseTest):
+
+    def test_search_by_keyword(self):
+        """
+        Tests if a user can search articles by keyword
+        """
+        self.add_article()
+        response = self.client.get(self.articles_url+"?keyword=the")
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
```
Comment (wasibani-roy) [inline authors/apps/articles/tests/test_search.py:19]: Hello @BagzieGracious good work on the search impressed by what you have done, however we where advised by the TTL to also test the response data

Comment (wasibani-roy) [inline authors/apps/articles/tests/test_search.py:19]: Also please add as to your reviewer's list there are only two people there

**Review by wasibani-roy — DISMISSED** [review]

File: authors/apps/articles/tests/test_search.py (hunk 1 above)
Comment (BagzieGracious) [reply authors/apps/articles/tests/test_search.py:19]: @wasibani-roy, thank you for the feedback. but i have fixed it

Comment (BagzieGracious) [reply authors/apps/articles/tests/test_search.py:19]: @wasibani-roy, Thank you for reminding me but i have added them now

File: authors/apps/articles/tests/test_search.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+class TestSearchArticle(ArticlesBaseTest):
+
+    def test_search_by_keyword(self):
+        """
+        Tests if a user can search articles by keyword
+        """
+        self.add_article()
+        response = self.client.get(self.articles_url+"?keyword=the")
+        print(response.data['articles'][0])
```
Comment (CryceTruly) [inline authors/apps/articles/tests/test_search.py]: Hey @BagzieGracious , thanks for the good work, However on line 18, I think you needed the print for debugging purposes. You could consider removing it if that's the case.

**Review by CryceTruly — CHANGES_REQUESTED** [review]

File: authors/apps/articles/views.py [hunk 3]
```diff
@@ -16,8 +18,55 @@
...
+    search_fields = (
+        'title',
+        'body',
+        'description',
+        'tagList',
+        'author__username',
+        'favorited_articles'
+    )
+
+    def get_queryset(self):
```
Comment (CryceTruly) [inline authors/apps/articles/views.py:32]: Thanks for the good work @BagzieGracious, however, the function `get_queryset` has a Cognitive Complexity of 6 (exceeds 5 allowed). You can consider refactoring.

**Review by CryceTruly — CHANGES_REQUESTED** [review]

**Review by joelethan — DISMISSED** [review]

**Review by joelethan — APPROVED** [review]

**Review by CryceTruly — APPROVED** [review]

**Review by Peace-Apple — APPROVED** [review]

**Pull request merged**